File: frontend/src/api/services/participant.js
import api from '@api';
import { ENDPOINTS } from '@api/endpoints';

/**
 * Participant only: Get current preferences (categories + budget).
 */
export async function getParticipantPreferences() {
  const { data } = await api.instance.get(ENDPOINTS.participant.preferences);
  return data;
}

/**
 * Participant only: Update preferences (categories + budget).
 * Accepts partial payload: { category_ids?: number[], budget?: string | number }
 */
export async function updateParticipantPreferences(patchData) {
  const { data } = await api.instance.patch(ENDPOINTS.participant.preferences, patchData);
  return data;
}

/**
 * Participant only: Create or update a swipe (like/dislike) on an event.
 * swipeData shape: { event_id: number, liked: boolean }
 */
export async function createSwipe(swipeData) {
  const { data } = await api.instance.post(ENDPOINTS.participant.swipes, swipeData);
  return data;
}

/**
 * Participant only: Get swipe history.
 */
export async function getSwipeHistory() {
  const { data } = await api.instance.get(ENDPOINTS.participant.swipeHistory);
  return data;
}

/**
 * Participant only: Get personalized event feed (AI-ranked).
 */
export async function getFeed() {
  const { data } = await api.instance.get(ENDPOINTS.participant.Feed);
  return data;
}

File: backend/api/views/participant.py
from drf_spectacular.utils import extend_schema, OpenApiResponse
from rest_framework.decorators import api_view, permission_classes, parser_classes
from rest_framework.parsers import JSONParser
from rest_framework.response import Response
from rest_framework import status
from ..utils import IsParticipantRole
from ..serializers.participant import (
    GetParticipantPreferencesSerializer,
    UpdateParticipantPreferencesSerializer,
    CreateSwipeSerializer, 
    GetSwipeHistorySerializer
)
@extend_schema(
    methods=['GET'],
    responses={200: GetParticipantPreferencesSerializer},
    description="Participant only:  Get current preferences (categories + budget).",
)
@extend_schema(
    methods=['PATCH'],
    request=UpdateParticipantPreferencesSerializer,
    responses={200: OpenApiResponse(description="Profile info updated successfully.")},
    description="Participant only: Update preferences (categories + budget).",
)
@api_view(["GET", "PATCH"])
@permission_classes([IsParticipantRole])
@parser_classes([JSONParser])
def manage_participant_preferences(request):
    """
    Participant only:
    - GET: Get current preferences (categories + budget).
    - PATCH: Update preferences (categories + budget).
    """
    if request.method == "GET":
        serializer = GetParticipantPreferencesSerializer(data={}, context={"participant": request.user})
        serializer.is_valid(raise_exception=True)
        return Response(serializer.data, status=status.HTTP_200_OK)

    elif request.method == "PATCH":
        serializer = UpdateParticipantPreferencesSerializer(data=request.data, context={"participant": request.user})
        serializer.is_valid(raise_exception=True)
        serializer.save()
        return Response({"detail": "Preferences updated successfully."}, status=status.HTTP_200_OK)
    

@extend_schema(
    methods=['POST'],
    request=CreateSwipeSerializer,
    responses={201: OpenApiResponse(description="Swipe recorded successfully."),},
    description="Participant only: Create a swipe (like/dislike) for an event.",
)
@api_view(["POST"])
@permission_classes([IsParticipantRole])
@parser_classes([JSONParser])
def create_swipe(request):
    """
    Participant only: Create a swipe (like/dislike) for an event.
    """
    serializer = CreateSwipeSerializer(data=request.data, context={"participant": request.user})
    serializer.is_valid(raise_exception=True)
    serializer.save()
    return Response({"detail": "Swipe recorded successfully."}, status=status.HTTP_201_CREATED)


@extend_schema(
    methods=['GET'],
    responses={200: GetSwipeHistorySerializer},
    description="Participant only:  Get current preferences (categories + budget).",
)
@api_view(["GET"])
@permission_classes([IsParticipantRole])
@parser_classes([JSONParser])
def get_swipe_history(request):
    """
    Participant only: Get swipe history.
    """
    serializer = GetSwipeHistorySerializer(data={}, context={"participant": request.user})
    serializer.is_valid(raise_exception=True)
    return Response(serializer.data, status=status.HTTP_200_OK)